### DISCUSSION TITLE
The Trouble with EXPOSE

### DISCUSSION THREAD

**Discussion by adesutherland** [adesutherland/CREXX#12] [Ideas]
In Classic REXX, EXPOSE is used to expose a variable used within a PROCEDURE to the call context. 

Another way to think about it is that it exposes global variables to the procedure - the same outcome but actually conceptually quite different. In any case as mechanism for global variables EXPOSE is flawed; where a PROCEDURE calls a PROCEDURE (two identical EXPOSEs would be needed), so it makes more sense to consider its motivation as to allow a "call by reference" mechanism.

The big issue is that the there is no way to use a different variable identifier for the variable as known in the procedure and its identifier outside the procedure. 

For example
  
     CALL PROC VAR_NAME

     ...
     
     PROC: PROCEDURE
         ARG VAR_NAME_USED_IN_PROC
     ...

VAR_NAME and VAR_NAME_USED_IN_PROC are different and decoupled.

However

     VAR = 1
     CALL PROC VAR_NAME
     SAY VAR
     ...
     
     PROC: PROCEDURE EXPOSE VAR
         ARG VAR_NAME_USED_IN_PROC
         VAR = Something
     ...

The problem is that 
1. The calling code must use the same id VAR which might be inconvenient
2. The calling code must take care ... what are the side effects of calling PROC?! Who would know VAR was changed without taking a lot of care ...

One way to address this is to deprecate EXPOSE and introduce a call by reference mechanisms; REXX level B uses pass by reference all the time (TBC - today it is not clear if a "copy on change" would be more natural for programmers) so perhaps expose is not needed ...

Another option is to stick with EXPOSE but somehow allow a **caller** to map the expose variables. REXX has mechanisms to make the EXPOSE more sophisticated but the issue is these are defined by the procedure definition and not by the caller. For example something like:

    CALL PROC ARG1, ARG2, EXPOSED1=MYVAR, EXPOSEDSTEM.=LOCALSTEM.

    ...
    PROC: PROCEDURE EXPOSE EXPOSED1 EXPOSEDSTEM.
       ARG A1, A2
       ...
       RETURN

Comments welcome!

**Comment by BobBolch** [comment]
>1.The calling code must use the same id VAR which might be inconvenient
>2.The calling code must take care ... what are the side effects of calling
PROC?! Who would know VAR was changed without taking a lot of care

Am I completely confused here? Isn't EXPOSE used only in internal
subroutines in the same source file as the calling program?
If so, then I don't understand what the inconvenience and confusion is.

Thanks/B

**Reply by adesutherland** [reply]
Bob - you make a strong point. On classic REXX it is indeed kind of one file per program. 

However as programs grow bigger clarity as to side effects are important. And after all Compiled REXX introduced `includes` (perhaps a gludge) that means that 3rd party includes could be used that are complex ... 

We know we have to implement Classic REXX -  no argument. But if we were reimagining REXX what could it look like ... that is what I am trying to work through

**Comment by BobBolch** [comment]
My development system for VM:Secure. VM:Backup, etc. used an INCLUDE
pseudo-statement for  common subroutine
packages, but we never thought we needed some different way of passing
modifiable arguments into a program.
That was for programming real world products with many thousands of lines
of REXX.  I just do not see any benefit
of your requirement. EXPOSE says "I want to not use any variables in my
calling program, EXCEPT for these I explicitly name."
It's a simple usable concept.

/B
